server/src/services/backup.service.ts

```diff
@@ -0,0 +1,178 @@
+    toDelete.push(...failedBackups);
+
+    for (const file of toDelete) {
+      await this.storageRepository.unlink(`${backupsFolder}/${file}`);
+    }
+    this.logger.debug(`Database Backup Cleanup Finished, deleted ${toDelete.length} backups`);
+  }
+
+  async handleBackupDatabase(): Promise<JobStatus> {
+    this.logger.debug(`Database Backup Started`);
+
+    const {
+      database: { config },
+    } = this.configRepository.getEnv();
+
+    const isUrlConnection = config.connectionType === 'url';
+    const databaseParams = isUrlConnection ? [config.url] : ['-U', config.username, '-h', config.host];
+
+    try {
+      await new Promise<void>((resolve, reject) => {
+        const pgdump = spawn(`pg_dumpall`, [...databaseParams, '--clean', '--if-exists'], {
+          env: { PATH: process.env.PATH, PGPASSWORD: isUrlConnection ? undefined : config.password },
+        });
+
+        const gzip = spawn(`gzip`, [], {
+          stdio: [pgdump.stdout, 'pipe', 'pipe'],
+        });
+
+        const backupFilePath = `${StorageCore.getBaseFolder(StorageFolder.BACKUPS)}/immich-db-backup-${Date.now()}.sql.gz.tmp`;
+
+        const fileStream = fs.createWriteStream(backupFilePath);
+
+        gzip.stdout.pipe(fileStream);
+
+        if (!pgdump.stderr || !pgdump.stdout) {
+          this.logger.error('Backup failed, could not spawn backup process');
+          reject('Backup failed, could not spawn backup process');
+          return;
+        }
+
+        if (!gzip.stderr || !gzip.stdout) {
+          this.logger.error('Backup failed, could not spawn gzip process');
+          reject('Backup failed, could not spawn gzip process');
+          return;
+        }
+
+        pgdump.on('error', (err) => {
+          this.logger.error('Backup failed with error', err);
+          reject(err);
+        });
+
+        gzip.on('error', (err) => {
+          this.logger.error('Gzip failed with error', err);
+          reject(err);
+        });
+
+        let pgdumpLogs = '';
+        let gzipLogs = '';
+
+        pgdump.stderr.on('data', (data) => (pgdumpLogs += data));
+        gzip.stderr.on('data', (data) => (gzipLogs += data));
+
+        pgdump.on('exit', (code) => {
+          if (code !== 0) {
+            this.logger.error(`Backup failed with code ${code}`);
+            reject(`Backup failed with code ${code}`);
+            this.logger.error(pgdumpLogs);
+            return;
+          }
+          if (pgdumpLogs) {
+            this.logger.debug(`pgdump_all logs\n${pgdumpLogs}`);
+          }
+        });
+
+        gzip.on('exit', (code) => {
+          if (code !== 0) {
+            this.logger.error(`Gzip failed with code ${code}`);
+            reject(`Gzip failed with code ${code}`);
+            this.logger.error(gzipLogs);
+            return;
+          }
+          if (pgdump.exitCode !== 0) {
+            this.logger.error(`Gzip exited with code 0 but pgdump exited with ${pgdump.exitCode}`);
+            return;
+          }
```

Comment:
this doesn't have a resolve/reject, does that imply an error event was thrown and you already returned something elsewhere?

Reply:
Yea, this is just handling the case that pgdump errored but gzip exits successfully, we have to handle that case differently.

Reply:
I mean where is the promise resolved or rejected in this scenario? In the `on('error')` event handler?